File: backend/app/Models/BankAccount.php
<?php

namespace App\Models;

use App\Models\Concerns\BelongsToBakery;
use App\Models\Concerns\RemembersLedgerTotal;
use App\Support\Money;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\Relation;

/**
 * A bank or cash account. The balance is derived from the opening figure
 * plus the movement ledger, never stored, so a transaction can never be
 * recorded without the balance following it.
 */
class BankAccount extends Model
{
    use BelongsToBakery, RemembersLedgerTotal;

    protected $fillable = [
        'title',
        'bank_name',
        'account_number',
        'card_number',
        'iban',
        'opening_balance',
        'is_default',
        'is_cash_box',
        'is_active',
        'note',
    ];

    protected function casts(): array
    {
        return [
            'opening_balance' => 'decimal:2',
            'is_default' => 'boolean',
            'is_cash_box' => 'boolean',
            'is_active' => 'boolean',
        ];
    }

    protected static function booted(): void
    {
        // Exactly one account can be the default, so setting a new one
        // clears the previous.
        static::saved(function (self $account) {
            if ($account->is_default) {
                static::where('id', '!=', $account->id)
                    ->where('is_default', true)
                    ->update(['is_default' => false]);
            }

            // And exactly one drawer. The migration that added the column
            // said so — «two tills is a question with no answer» — and
            // then only the default was ever held to it, so a second one
            // could be flagged and `cashBox()` would answer with whichever
            // came first by id.
            //
            // Separate from the default on purpose: a one-account shop
            // keeps its cash and its takings in the same place, and
            // neither flag should push the other off.
            if ($account->is_cash_box) {
                static::where('id', '!=', $account->id)
                    ->where('is_cash_box', true)
                    ->update(['is_cash_box' => false]);
            }
        });
    }

    public function transactions()
    {
        return $this->hasMany(BankTransaction::class);
    }

    /**
     * Loads the signed ledger total with the rows, so a list of accounts
     * costs one question rather than one per account.
     */
    public function scopeWithBalance(Builder $query): Builder
    {
        return $query
            ->withSum(['transactions as ledger_in' => fn ($q) => $q->where('direction', 'in')], 'amount')
            ->withSum(['transactions as ledger_out' => fn ($q) => $q->where('direction', 'out')], 'amount');
    }

    public function getBalanceAttribute(): float
    {
        // `withSum` cannot add the opening figure, so the preloaded column
        // is the ledger alone and the opening balance goes on here either way.
        if (array_key_exists('ledger_in', $this->attributes) && $this->memoTakenAt === null) {
            return $this->rememberLedgerTotal(
                fn () => round(
                    (float) $this->opening_balance
                    + (float) $this->attributes['ledger_in']
                    - (float) $this->attributes['ledger_out'],
                    2
                )
            );
        }

        return $this->rememberLedgerTotal(function () {
            // One pass over the ledger, signed, rather than one sum each
            // way: the two answers were the same table read twice.
            $net = (float) $this->transactions()
                ->selectRaw("coalesce(sum(case when direction = 'in' then amount else -amount end), 0) as net")
                ->value('net');

            return round((float) $this->opening_balance + $net, 2);
        });
    }

    public function getBalanceFormattedAttribute(): string
    {
        return Money::format($this->balance);
    }

    /** A negative balance means the account is overdrawn. */
    public function getIsOverdrawnAttribute(): bool
    {
        return $this->balance < 0;
    }

    public function getLabelAttribute(): string
    {
        return $this->bank_name
            ? "{$this->title} — {$this->bank_name}"
            : $this->title;
    }

    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    /** Records a movement and returns it. */
    public function record(
        string $direction,
        float $amount,
        string $reason = 'manual',
        ?int $userId = null,
        ?Model $source = null,
        ?string $note = null,
        $occurredOn = null,
    ): BankTransaction {
        return $this->transactions()->create([
            'user_id' => $userId,
            'direction' => $direction,
            'amount' => $amount,
            'reason' => $reason,
            'source_type' => $source ? Relation::getMorphAlias($source::class) : null,
            'source_id' => $source?->getKey(),
            'occurred_on' => $occurredOn ?? now(),
            'note' => $note,
        ]);
    }

    /** The account money defaults to, or the only active one. */
    /**
     * The drawer, as opposed to a bank.
     *
     * Found by its flag rather than its name: matching on "صندوق نقد"
     * would hold until someone renamed it on a screen that invites
     * renaming, and cash would quietly stop being recorded again.
     */
    public static function cashBox(): ?self
    {
        return static::active()->where('is_cash_box', true)->first();
    }

    public static function defaultAccount(): ?self
    {
        return static::active()->where('is_default', true)->first()
            ?? static::active()->first();
    }

    /**
     * Where money taken on the reader lands — «درامد کارتخوان فقط بره حساب
     * سفید».
     *
     * The default account, unless that is the drawer. A shop can flag one
     * account as both, and then card takings would be booked as cash in
     * hand: the drawer reads high by every card sale, and the difference is
     * only found by somebody counting the notes and not finding them.
     *
     * Nothing rather than the drawer when there is no bank at all. A card
     * payment that posts nowhere is a gap somebody can still go and look
     * for; one sitting in the till is a figure that looks right and is not.
     */
    public static function cardAccount(): ?self
    {
        $default = static::active()->where('is_default', true)->first();

        if ($default && ! $default->is_cash_box) {
            return $default;
        }

        // No usable default. One bank and there is nothing to decide, so
        // the money goes where it can only have gone. Several and this
        // stops: picking the lowest id would be a guess, and a guess about
        // which account took the money is worse than a gap somebody can
        // still go and look for.
        $banks = static::active()->where('is_cash_box', false)->limit(2)->get();

        return $banks->count() === 1 ? $banks->first() : null;
    }
}

File: mobile-app/lib/models/bank_account.dart
/// A bank account the shop banks its card takings into.
///
/// Balances are derived on the server from the account's own ledger, so the
/// app only ever reads them — there is no figure here to keep in step.
class BankAccount {
  const BankAccount({
    required this.id,
    required this.title,
    required this.balance,
    required this.balanceFormatted,
    this.label,
    this.bankName,
    this.isDefault = false,
    this.isActive = true,
    this.isOverdrawn = false,
  });

  final int id;
  final String title;

  /// The shop's own name for it, plus the bank, when both are recorded.
  final String? label;

  final String? bankName;

  final double balance;

  /// Already carries the currency the shop displays in, so the app does not
  /// have to know whether it is toman or rial.
  final String balanceFormatted;

  final bool isDefault;
  final bool isActive;
  final bool isOverdrawn;

  static double _toDouble(dynamic value) => switch (value) {
        num n => n.toDouble(),
        String s => double.tryParse(s) ?? 0,
        _ => 0,
      };

  static bool _toBool(dynamic value) => value == true || value == 1 || value == '1';

  factory BankAccount.fromJson(Map<String, dynamic> json) => BankAccount(
        id: (json['id'] as num?)?.toInt() ?? 0,
        title: json['title'] as String? ?? '',
        label: json['label'] as String?,
        bankName: json['bank_name'] as String?,
        balance: _toDouble(json['balance']),
        balanceFormatted: json['balance_formatted'] as String? ?? '',
        isDefault: _toBool(json['is_default']),
        isActive: _toBool(json['is_active']),
        isOverdrawn: _toBool(json['is_overdrawn']),
      );
}

/// Every account, with what they come to together.
class BankBalances {
  const BankBalances({
    required this.accounts,
    required this.totalFormatted,
    required this.total,
  });

  final List<BankAccount> accounts;
  final double total;
  final String totalFormatted;

  /// The total counts only accounts still in use, the way the server sums
  /// it — a closed account's balance is not money the shop can spend.
  factory BankBalances.fromJson(Map<String, dynamic> json) => BankBalances(
        accounts: ((json['accounts'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(BankAccount.fromJson)
            .toList(),
        total: BankAccount._toDouble(json['total_balance']),
        totalFormatted: json['total_balance_formatted'] as String? ?? '',
      );

  bool get isEmpty => accounts.isEmpty;
}
